File: Polymorphism/Polymorphism.java
package Polymorphism;

class Calculator {
    // 1. Compile-time Polymorphism (Method Overloading)
    public int add(int a, int b) {
        return a + b;
    }

    public double add(double a, double b) {
        return a + b;
    }
}

class Animal {
    public void makeSound() {
        System.out.println("Animal makes a sound");
    }
}

class Dog extends Animal {
    // 2. Run-time Polymorphism (Method Overriding)
    @Override
    public void makeSound() {
        System.out.println("Dog barks");
    }
}

public class Polymorphism {
    public static void main(String[] args) {
        // Test Overloading
        Calculator calc = new Calculator();
        System.out.println("Sum (int): " + calc.add(5, 10));
        System.out.println("Sum (double): " + calc.add(5.5, 10.5));

        // Test Overriding
        Animal myAnimal = new Dog(); // Upcasting
        myAnimal.makeSound(); // Calls Dog's method at runtime
    }
}

File: Polymorphism/Polymorphism.cpp
#include <iostream>
using namespace std;

class Calculator {
public:
    // 1. Compile-time Polymorphism (Function Overloading)
    int add(int a, int b) {
        return a + b;
    }

    double add(double a, double b) {
        return a + b;
    }
};

class Animal {
public:
    // Virtual function for Run-time Polymorphism
    virtual void makeSound() {
        cout << "Animal makes a sound" << endl;
    }
};

class Dog : public Animal {
public:
    // 2. Run-time Polymorphism (Method Overriding)
    void makeSound() override {
        cout << "Dog barks" << endl;
    }
};

int main() {
    // Test Overloading
    Calculator calc;
    cout << "Sum (int): " << calc.add(5, 10) << endl;
    cout << "Sum (double): " << calc.add(5.5, 10.5) << endl;

    // Test Overriding
    Animal* myAnimal = new Dog(); // Upcasting using pointer
    myAnimal->makeSound(); // Calls Dog's method at runtime

    delete myAnimal; // Clean up memory
    return 0;
}
